python/graphstorm/trainer/glem_np_trainer.py

```diff
@@ -256,41 +256,48 @@ def _prepare_batch(input_nodes, seeds, blocks, is_labeled=True):
+                                      use_mini_batch_infer, total_steps, return_proba=False)
 
                 if self.evaluator.do_early_stop(val_score):
                     self.early_stop = True
 
+            # In every save_model_frequency iterations, check to save the top k models. 
+            # If has validation score, will save the best top k. If no validation, will
+            # either save the last k model or all models depends on the setting of top k.
             if save_model_frequency > 0 and \
                 total_steps % save_model_frequency == 0 and \
                 total_steps != 0:
-                if self.evaluator is None or val_score is not None:
+                if self.can_do_model_eval(val_loader):
+                    # for model saving, force to do evaluation if can
+                    val_score = self.eval(model.module, val_loader, test_loader,
+                                          use_mini_batch_infer, total_steps, return_proba=False)
                     # We will save the best model when
-                    # 1. There is no evaluation, we will keep the
-                    #    latest K models.
-                    # 2. There is evaluaiton, we need to follow the
-                    #    guidance of validation score.
-                    self.save_topk_models(model, epoch, i, val_score, save_model_path)
+                    # 1. If not do evaluation, we will keep the latest K models.
+                    # 2. If do evaluaiton, we need to follow the guidance of validation score.
+                self.save_topk_models(model, epoch, i, val_score, save_model_path)
+
             rt_profiler.record('train_eval')
             # early_stop, exit current interation.
             if self.early_stop is True:
                 break
 
-        # end of an epoch
+        # ------- end of an epoch -------
+
         barrier()
 
         val_score = None
-        if self.evaluator is not None and self.evaluator.do_eval(total_steps, epoch_end=True):
+        # do evaluation and mode saving after each epoch if can
+        if self.can_do_model_eval(val_loader):
             val_score = self.eval(model.module, val_loader, test_loader,
-                                    use_mini_batch_infer, total_steps, return_proba=False)
+                                  use_mini_batch_infer, total_steps, return_proba=False)
             if self.evaluator.do_early_stop(val_score):
                 self.early_stop = True
+
         # After each epoch, check to save the top k models. If has validation score, will save
         # the best top k. But if no validation, will either save the last k model or all models
         # depends on the setting of top k. To show this is after epoch save, set the iteration
         # to be None, so that we can have a determistic model folder name for testing and debug.
         self.save_topk_models(model, epoch, None, val_score, save_model_path)
+        barrier()
```

Comment:
Why is this barrier here? What's the potential issue of workers don't sync at this point?

Reply:
There is one case. When a trainer (local_rank != 0) completes all epoch and start to save model, particularly saving a large number of learnable embeddings. If not barrier(), the main process (local_rank = 0) complete its saving and start to kill all other trainers. Then some saving process may not finish all saving.

Reply:
Can we add a comment for this? Barriers can cause delays and should be motivated

Reply:
updated with a comment line to explain the reason.